handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,492 @@
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assume.assumeTrue;
+
+public class OcspTest {
+
+    @BeforeClass
+    public static void checkOcspSupported() {
+        assumeTrue(OpenSsl.isOcspSupported());
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testJdkClientEnableOcsp() throws Exception {
+        SslContextBuilder.forClient()
+                .sslProvider(SslProvider.JDK)
+                .enableOcsp(true)
+                .build();
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testJdkServerEnableOcsp() throws Exception {
+        SelfSignedCertificate ssc = new SelfSignedCertificate();
+        try {
+            SslContextBuilder.forServer(ssc.certificate(), ssc.privateKey())
+                    .sslProvider(SslProvider.JDK)
+                    .enableOcsp(true)
+                    .build();
+        } finally {
+            ssc.delete();
+        }
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testClientOcspNotEnabledOpenSsl() throws Exception {
+        testClientOcspNotEnabled(SslProvider.OPENSSL);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testClientOcspNotEnabledOpenSslRefCnt() throws Exception {
+        testClientOcspNotEnabled(SslProvider.OPENSSL_REFCNT);
+    }
+
+    private void testClientOcspNotEnabled(SslProvider sslProvider) throws Exception {
+        SslContext context = SslContextBuilder.forClient()
+                .sslProvider(sslProvider)
+                .build();
+        try {
+            SslHandler sslHandler = context.newHandler(ByteBufAllocator.DEFAULT);
+            ReferenceCountedOpenSslEngine engine = (ReferenceCountedOpenSslEngine) sslHandler.engine();
+            engine.getOcspResponse();
+        } finally {
+            ReferenceCountUtil.release(context);
+        }
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testServerOcspNotEnabledOpenSsl() throws Exception {
+        testServerOcspNotEnabled(SslProvider.OPENSSL);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testServerOcspNotEnabledOpenSslRefCnt() throws Exception {
+        testServerOcspNotEnabled(SslProvider.OPENSSL_REFCNT);
+    }
+
+    private void testServerOcspNotEnabled(SslProvider sslProvider) throws Exception {
+        SelfSignedCertificate ssc = new SelfSignedCertificate();
+        try {
+            SslContext context = SslContextBuilder.forServer(ssc.certificate(), ssc.privateKey())
+                    .sslProvider(sslProvider)
+                    .build();
+            try {
+                SslHandler sslHandler = context.newHandler(ByteBufAllocator.DEFAULT);
+                ReferenceCountedOpenSslEngine engine = (ReferenceCountedOpenSslEngine) sslHandler.engine();
+                engine.setOcspResponse(new byte[] { 1, 2, 3 });
```

Comment:
the `engine` must be released at some point. `SslHandler` typically takes ownership, but it must be interred into a pipeline ignorer to do so.